### DISCUSSION TITLE
Literal message is missing as per PGP standard

### DISCUSSION THREAD

**Discussion by varsha123** [openpgpjs/openpgpjs#1524] [Q&A]
Hi Team @larabr 

Our client facing error as "Literal message is missing as per PGP standard" while decrypting and verifying File signed and encrypted with openpgpJS .

Plz check below code we are using and suggest what's wrong

```js
 const publicKeyArmored = fs.readFileSync('keys/#############.asc', { encoding: 'utf8' });

  const publicKey = await openpgp.readKey({ armoredKey: publicKeyArmored });

  const privateKeyArmored = fs.readFileSync('keys/#####-private-key.key', { encoding: 'utf8'}); // encrypted private key
  
  const passphrase = `#################`; // what the private key is encrypted with

  const publicKey = await openpgp.readKey({ armoredKey: publicKeyArmored });

  const privateKey = await openpgp.decryptKey({
    privateKey: await openpgp.readPrivateKey({ armoredKey: privateKeyArmored }),
    passphrase
  });
  /**============================================================================== */
  
  const message = fs.readFileSync("keys/##################abc.txt", { encoding: 'utf-8' });
  const unsignedMessage = await openpgp.createMessage({
    text: message,
  });
  
  const encrypted = await openpgp.encrypt({
    message: unsignedMessage,
    encryptionKeys: publicKey, // public key
    signingKeys: privateKey,// tp private key
    config: { preferredCompressionAlgorithm: openpgp.enums.compression.zlib } 
  });
  console.log('encrypted', encrypted);
  fs.writeFileSync("one_step_encrypted_request.sig", encrypted);
```

Thanks
Varsha

**Comment by twiss** [comment]
Hi, which software (and version) are you using to decrypt the message?

We certainly create a literal data packet, it's not possible to create a message without it.

Perhaps the receiving software is confused by the compressed data packet, and expects a literal data packet instead? You could try removing the compression config to test that.

**Reply by varsha123** [reply]
Hi @twiss 

Using bouncycastle library to decrypt the message.

Also they informed us below details:

please find the PGP details at our side. Also, please enable the compression flag at your side. Thanks!
 
1.1.1      Digital Signature
Type                           :             RSA
Key Size                      :             2048 bits
Hashing Algorithm   :             SHA-2
 
1.1.2      Data encryption
Type                           :             AES -128 or AES-256
Key Size                      :             256 bits

For compression we have used :  config: { preferredCompressionAlgorithm: openpgp.enums.compression.zlib } 

Earlier we had shred without compression but that also had same error and after encryption also there is same error ,
how can we remove default config for compression.?

Please let us know is it our above code  following this config

Thanks
Varsha

**Reply by twiss** [reply]
By default we don't use compression.

One other stab in the dark: perhaps you are trying to parse it as an (unencrypted) signed message, rather than an encrypted message? That might be why they're expecting a literal data packet.

In general, you might want to refer to Bouncycastle regarding how to decrypt such a message, and/or why that library doesn't accept it.

**Reply by varsha123** [reply]
Hi @twiss  and team

Client is able to verify and decrypt if we go for 2 step  like sign using openpgp.sign method and encrypt using openpgp.encrypt method..

But if we go with one step sign and encryption using below code , they are not getting Literal message and getting error :
Literal message is missing ..   

**Can you please confirm, is it openpgpJS supporting one step sign and encryption**
At client end they are using one step for decrypt-Verify (which is decrypting first and then verifying message).

```js
 const message = fs.readFileSync("keys/##################abc.txt", { encoding: 'utf-8' });
  const unsignedMessage = await openpgp.createMessage({
    text: message,
  });
  
  const encrypted = await openpgp.encrypt({
    message: unsignedMessage,
    encryptionKeys: publicKey, // public key
    signingKeys: privateKey,// tp private key
    config: { preferredCompressionAlgorithm: openpgp.enums.compression.zlib } 
  });
  console.log('encrypted', encrypted);
  fs.writeFileSync("one_step_encrypted_request.sig", encrypted);
```

Thanks
Varsha

**Reply by twiss** [reply]
Yes, we support it, it should work. Just the fact that you're writing it to a `.sig` file is a bit strange but it shouldn't have an effect I think, but that depends on how BouncyCastle is processing the message.

**Reply by varsha123** [reply]
ok Thanks for the update, can you please confirm, what's sequence of  (one step sign and encryption)?
Is it like first signed and then same signed message getting encrypted..

What should be file extension and can we try without extension?

Thanks
Varsha

**Reply by twiss** [reply]
Yes, the message is first signed and then the signed message is encrypted. There is no standard for extensions, except perhaps .pgp or .asc for armored files. But I don't think that's the issue, unless BouncyCastle is trying to read the file as a signature, specifically.

**Comment by varsha123** [comment]
Hi @twiss why Literal data is missing after sign when use one step sign-encrypt?
how can we verify this at our end.

Thanks
Varsha

**Reply by twiss** [reply]
The literal data packet is inside the encrypted message when it's signed and encrypted.
